### PR TITLE
[SPARK-26923][SQL][R] Refactor ArrowRRunner and RRunner to share one BaseRRunner

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to have one base R runner. 

In the high level,

Previously, it had `ArrowRRunner` and it inherited `RRunner`:

```
└── RRunner
    └── ArrowRRunner
```

After this PR, now it has a `BaseRRunner`, and `ArrowRRunner` and `RRunner` inherit `BaseRRunner`:

```
└── BaseRRunner
    ├── ArrowRRunner
    └── RRunner
```

This way is consistent with Python's.

In more details, see below:

```scala
class BaseRRunner[IN, OUT] {

  def compute: Iterator[OUT] = {
    ...
    newWriterThread(...).start()
    ...
    newReaderIterator(...)
    ...
  }

  // Make a thread that writes data from JVM to R process
  abstract protected def newWriterThread(..., iter: Iterator[IN], ...): WriterThread

  // Make an iterator that reads data from the R process to JVM
  abstract protected def newReaderIterator(...): ReaderIterator

  abstract class WriterThread(..., iter: Iterator[IN], ...) extends Thread {
    override def run(): Unit {
      ...
      writeIteratorToStream(...)
      ...
    }

    // Actually writing logic to the socket stream.
    abstract protected def writeIteratorToStream(dataOut: DataOutputStream): Unit
  }

  abstract class ReaderIterator extends Iterator[OUT] {
    override def hasNext(): Boolean = {
      ...
      read(...)
      ...
    }

    override def next(): OUT = {
      ...
      hasNext()
      ...    
    }

    // Actually reading logic from the socket stream.
    abstract protected def read(...): OUT
  }
}
```

```scala
case [Arrow]RRunner extends BaseRRunner {
  override def newWriterThread(...) {
    new WriterThread(...) {
      override def writeIteratorToStream(...) {
        ...
      }
    }
  }

  override def newReaderIterator(...) {
    new ReaderIterator(...) {
      override def read(...) {
        ...
      }
    }
  }
}
```

## How was this patch tested?

Manually tested and existing tests should cover.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
I need to clean up few more times. Let me leave cc anyway. @felixcheung, @BryanCutler, @ueshin, @viirya

File: core/src/main/scala/org/apache/spark/api/r/RRunner.scala [hunk 1]
```diff
@@ -44,380 +35,149 @@ private[spark] class RRunner[U](
...
+                  init,
+                  broadcast,
+                  input,
+                  compute,
+                  output,
+                  boot + init + broadcast + input + compute + output))
+              read()
+            case length if length > 0 =>
+              readData(length).asInstanceOf[OUT]
+            case length if length == 0 =>
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: I tried to not add or remove existing logics but this logic was added to reuse `ReaderIterator`. The end of stream should be marked explicitly (this is consistent with Python side).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 2]
```diff
@@ -546,14 +548,23 @@ case class FlatMapGroupsInRWithArrowExec(
...
+      val groupedByRKey: Iterator[Iterator[InternalRow]] =
+        grouped.map { case (key, rowIter) => rowIter }
+
+      val runner = new ArrowRRunner(func, packageNames, broadcastVars, inputSchema,
+        SQLConf.get.sessionLocalTimeZone, RRunnerModes.DATAFRAME_GAPPLY) {
+        protected override def bufferedWrite(
+            dataOut: DataOutputStream)(writeFunc: ByteArrayOutputStream => Unit): Unit = {
+          super.bufferedWrite(dataOut)(writeFunc)
+          // Don't forget we're sending keys additionally.
+          keys.foreach(dataOut.write)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: This logic is changed too. I tried to use `ArrowRRunner` to only send Arrow batches, and send keys separately here.

**Comment by HyukjinKwon** [comment]
adding @shivaram  too

**Comment by HyukjinKwon** [comment]
This is ready for a look.

File: core/src/main/scala/org/apache/spark/api/r/BaseRRunner.scala [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+    // The stdout/stderr is shared by multiple tasks, because we use one daemon
+    // to launch child process as worker.
+    val errThread = BaseRRunner.createRWorker(listenPort)
+
+    // We use two sockets to separate input and output, then it's easy to manage
+    // the lifecycle of them to avoid deadlock.
+    // TODO: optimize it to use one socket
+
+    // the socket used to send out the input of task
+    serverSocket.setSoTimeout(10000)
```
Comment (felixcheung) [inline core/src/main/scala/org/apache/spark/api/r/BaseRRunner.scala]: I think there's a JIRA on the connection timeout not being configurable...

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/r/BaseRRunner.scala]: D'oh, I only checked R side. Let me check Scala side as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala (hunk 2 above)
Comment (felixcheung) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: so we expect `keys` is small?

Comment (felixcheung) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: It's a bit hard for me to visualize... isnt each row associated with a key, which could be different? if I understand before it's a pairing for `(key, row)` or `(key, rowIter)` - how does it handle this now?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala [hunk 4]
```diff
@@ -202,14 +175,17 @@ class ArrowRRunner(
               // Likewise, there looks no way to send each batch in streaming format via socket
               // connection. See ARROW-4512.
               // So, it reads the whole Arrow streaming-formatted binary at once for now.
-              val in = new ByteArrayReadableSeekableByteChannel(readByteArrayData(length))
+              val buffer = new Array[Byte](length)
```
Comment (felixcheung) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala]: why replace `readByteArrayData`?

File: core/src/main/scala/org/apache/spark/api/r/RRunner.scala (hunk 1 above)
Comment (felixcheung) [reply core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: I see this new case - how can we test this more?

**Review by felixcheung — COMMENTED** [review]
I checked all the moved code. a few comments

File: core/src/main/scala/org/apache/spark/api/r/RRunner.scala (hunk 1 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: This is actually being tested by existing tests in R. (I am going to explain from scratch to completely make sure we're synced on this).

Here is when the data format is sent from R process to JVM via socket.


```
----------------------------
| ... # Output data        |
----------------------------
| -1 # Timing data start   |
| ... # Computed time, etc.|
----------------------------
| 0 # End of stream        |
----------------------------
```

**Before:**

Previously, RRunner's iterator in this code path stopped iteration when `read()` returns `null`. So, `readData` returned `null` when it's `0`.

**After:**

Now, RRunner needs to mark the end of stream explicitly via setting `eos` to `true` (This prevents when the returned value of `read()` is `null` and it has some meanings about it).

So, in other words, it works if we mark `eos` to `true` at everywhere it returned `null` in the previous RRunner, 

https://github.com/apache/spark/blob/8126d09fb5b969c1e293f1f8c41bec35357f74b5/core/src/main/scala/org/apache/spark/api/r/RRunner.scala#L290

https://github.com/apache/spark/blob/8126d09fb5b969c1e293f1f8c41bec35357f74b5/core/src/main/scala/org/apache/spark/api/r/RRunner.scala#L282

https://github.com/apache/spark/blob/8126d09fb5b969c1e293f1f8c41bec35357f74b5/core/src/main/scala/org/apache/spark/api/r/RRunner.scala#L272

However, explicitly marking the end of stream when 0 is read at one place should be better. So I decided to do it as so.

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: In case or ArrowRRunner, it behaves same as previous.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: Yea, keys are expected to be small. I am also going to try to explain from scratch to make sure we're synced.

Input iterator:

```
[a, 1]
[b, 2]
[b, 3]
```

Output iterator:

```
([a], Iterator([a, 1])
([b], Iterator([b, 2], [b, 3])
```

Here, I am doing:

```
next() call => ([a], Iterator([a, 1])
  keeps the key 'a'
  converts Iterator([a, 1]) to Arrow batch
  buffers the Arrow batch
next() call => ([b], Iterator([b, 2], [b, 3])
  keeps the key 'a'
  converts Iterator([b, 2], [b, 3]) to Arrow batch
  buffers the Arrow batch
...

Write buffered all Arrow batches
Write kept all the keys.
```

I tried to only leave Arrow conversion logics in `ArrowRRunner` (which is shared for both `dapply` and `gapply`), and keeping/writing keys logic is here separate.

(BTW, I tried to avoid those refactorings cus I know it's difficult to review such PRs).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala (hunk 4 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala]: Just realised that `readByteArrayData` is only two lines, and thought it's better to put all reading related logics from RRunner into `RRunner`'s `ReaderIterator` in `newReaderIterator`.

File: core/src/main/scala/org/apache/spark/api/r/BaseRRunner.scala (hunk 3 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/r/BaseRRunner.scala]: Actually, this one is something already existing before in RRunner side, and there are multiple instances in Python side as well. The default value of `spark.r.backendConnectionTimeout` is 6000. So, if we change this, this could cause potential behaviour change. Let me leave those out in this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala (hunk 2 above)
Comment (felixcheung) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: does `[a]` `[b]` mean in your illustration an array of 1?

File: core/src/main/scala/org/apache/spark/api/r/BaseRRunner.scala (hunk 3 above)
Comment (felixcheung) [reply core/src/main/scala/org/apache/spark/api/r/BaseRRunner.scala]: yes I think it's https://issues.apache.org/jira/browse/SPARK-12609

**Comment by felixcheung** [comment]
I reviewed then https://github.com/apache/spark/pull/23977#pullrequestreview-211607985

only one new comment https://github.com/apache/spark/pull/23977#discussion_r263988629

**Review by srowen — COMMENTED** [review]
Generally looks good as a refactoring

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: Oh, right. The types look confusing.

Input iterator:

`Iterator[InternalRow]`

Output Iterator

`Iterator[(InternalRow, Iterator[InternalRow])]`

```bash
next() call => ([a], Iterator([a, 1])             # (InternalRow, Iterator[InternalRow])
  1. converts the key 'a' into R-readable bytes   # 'a' becomes InternalRow -> Array[Byte] (see [1])
  2. keeps the converted key in `keys`            # 'keys` is mutable.ArrayBuffer[Array[Byte] (see [1])
  3. converts Iterator([a, 1]) to Arrow batch     # Iterator[InternalRow] -> Array[Byte]
  4. buffers the Arrow batch                      # buffer: Array[Byte] + Arrow batch: Array[Byte]
...
Write buffered all Arrow batches                  # See [2]
Write all the keys.                               # See [3]
```

[1] https://github.com/apache/spark/pull/23977/files#diff-a1240a7ba20d0e027d23a690a770bf44R557
[2] https://github.com/apache/spark/pull/23977/files#diff-a0b6a11cc2e2299455c795fe3c96b823R71
[3] https://github.com/apache/spark/pull/23977/files#diff-a1240a7ba20d0e027d23a690a770bf44R567

**Comment by HyukjinKwon** [comment]
retest this please

**Review by BryanCutler — COMMENTED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Merged to master.

Thank you guys.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala (hunk 2 above)
Comment (felixcheung) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: ok thanks for the explanation
